cylc/flow/task_pool.py

```diff
@@ -318,73 +318,45 @@ def compute_runahead(self, force=False) -> bool:
-                ]
-                # Drop points beyond the limit.
-                points = sorted(points)[:ilimit + 1]
-                base_point = min(points)
-
-            else:
-                # Start at first point in each sequence.
-                # (After workflow start point - sequence may begin earlier).
-                points = [
-                    point
-                    for point in {
-                        seq.get_first_point(self.config.start_point)
-                        for seq in self.config.sequences
-                    }
-                    if point is not None
-                ]
-                base_point = min(points)
-                # Drop points beyond the limit.
-                points = [
-                    point
-                    for point in points
-                    if point <= base_point + limit
-                ]
-
+            # Find the earliest sequence point beyond the workflow start point.
+            points = [
+                point
+                for point in {
+                    seq.get_first_point(self.config.start_point)
+                    for seq in self.config.sequences
+                }
+                if point is not None
+            ]
+            base_point = min(points)
         else:
             # Find the earliest point with unfinished tasks.
             for point, itasks in sorted(self.get_tasks_by_point().items()):
-                if (
-                    points  # got the limit already so this point too
-                    or any(
-                        not itask.state(
-                            TASK_STATUS_FAILED,
-                            TASK_STATUS_SUCCEEDED,
-                            TASK_STATUS_EXPIRED
-                        )
-                        or (
-                            # For Cylc 7 back-compat, ignore incomplete tasks.
-                            # (Success is required in back-compat mode, so
-                            # failedtasks end up as incomplete; and Cylc 7
-                            # ignores failed tasks in computing the limit).
-                            itask.state.outputs.is_incomplete()
-                            and not cylc.flow.flags.cylc7_back_compat
-                        )
-                        for itask in itasks
+                if any(
+                    not itask.state(
+                        TASK_STATUS_FAILED,
+                        TASK_STATUS_SUCCEEDED,
+                        TASK_STATUS_EXPIRED
+                    ) or (
+                        # For Cylc 7 back-compat, ignore incomplete tasks.
+                        # (Success is required in back-compat mode, so
+                        # failed tasks end up as incomplete; and Cylc 7
+                        # ignores failed tasks in computing the limit).
+                        itask.state.outputs.is_incomplete()
+                        and not cylc.flow.flags.cylc7_back_compat
                     )
```

Comment:
The "Cylc 7 back-compat" comment is actually attached to the Cylc 8 logic.

Moreover, the Cylc 7 logic can interfere with the Cylc 8 logic because the `cylc.flow.flags.cylc7_back_compat` statement is only on one branch of the `or` (potential for future bugs?).

Would suggest either:

```python
if any(
    (
        not cylc.flow.flags.cylc7_back_compat
        and itask.state.outputs.is_incomplete()
    ) or (
        # For Cylc 7 back-compat, ignore incomplete tasks.
        # (Success is required in back-compat mode, so
        # failed tasks end up as incomplete; and Cylc 7
        # ignores failed tasks in computing the limit).
        cylc.flow.flags.cylc7_back_compat
        and not itask.state(
            TASK_STATUS_FAILED,
            TASK_STATUS_SUCCEEDED,
            TASK_STATUS_EXPIRED
        )
    )
    for itask in itasks
)
```

Or the more efficient:

```python
if cylc.flow.flags.cylc7_back_compat:
    if any(itask.state.outputs.is_incomplete() for itask in itasks)
else:
    # For Cylc 7 back-compat, ignore incomplete tasks.
    # (Success is required in back-compat mode, so
    # failed tasks end up as incomplete; and Cylc 7
    # ignores failed tasks in computing the limit).
    if any(
        not itask.state(
            TASK_STATUS_FAILED,
            TASK_STATUS_SUCCEEDED,
            TASK_STATUS_EXPIRED
        )
        for itask in itasks
    )
)
```

Reply:
Good call.